onnxruntime/core/framework/partial_executor.cc

```diff
@@ -0,0 +1,497 @@
+    }
+#endif
+
+    auto p_op_kernel = session_state.GetKernel(node_index);
+    if (p_op_kernel->KernelDef().OpName() == "YieldOp") {
+      // Do not execute YieldOp (it is an no-op anyways).
+      // Decrement the reference count of tensors that are not needed beyond this point.
+      // REVEIW(codemzs): The current model assumes the intermediate tensors that are exported
+      // as graph outputs are owned by ORT, the risk of caller freeing the tensor or manipulating tensor 
+      // memory lingers while the tensor is used downstream after the export.
+      VLOGS(logger, 1) << "Releasing node ML values.";
+      ORT_RETURN_IF_ERROR(ReleaseNodeMLValues(frame, seq_exec_plan, node_exec_plan, logger));
+      continue;
+    }
+    // if a kernel has been added in the session state, it better be NON-null.
+    if (p_op_kernel == nullptr)
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Got nullptr from GetKernel for node: ",
+                             node.Name());
+
+#ifdef ONNXRUNTIME_ENABLE_INSTRUMENT
+    LARGE_INTEGER kernel_start;
+    QueryPerformanceCounter(&kernel_start);
+#endif
+    // construct OpKernelContext
+    // TODO: log kernel inputs?
+    OpKernelContextInternal op_kernel_context(session_state, frame, *p_op_kernel, logger, terminate_flag_);
+    // TODO: log kernel outputs?
+    if (is_profiler_enabled) {
+      sync_time_begin = session_state.Profiler().Now();
+    }
+
+    // sync before compute
+    int queue_id = p_op_kernel->KernelDef().ExecQueueId();
+    if (seq_exec_plan.NodeHasFence(node_index)) {
+      for (int input_index = 0; input_index < op_kernel_context.InputCount(); ++input_index) {
+        Fence_t fence = op_kernel_context.InputFence(input_index);
+        if (fence) {
+          auto execution_provider_type = p_op_kernel->Node().GetExecutionProviderType();
+          if (OrtMemTypeCPUInput == p_op_kernel->KernelDef().InputMemoryType(input_index)) {
+            execution_provider_type = kCpuExecutionProvider;
+          }
+          fence->BeforeUsingAsInput(execution_provider_type, queue_id);
+        }
+      }
+
+      for (int input_index = 0; input_index < op_kernel_context.ImplicitInputCount(); ++input_index) {
+        Fence_t fence = op_kernel_context.ImplicitInputFence(input_index);
+        if (fence) {
+          auto execution_provider_type = p_op_kernel->Node().GetExecutionProviderType();
+          if (OrtMemTypeCPUInput == p_op_kernel->KernelDef().InputMemoryType(input_index)) {
+            execution_provider_type = kCpuExecutionProvider;
+          }
+          fence->BeforeUsingAsInput(execution_provider_type, queue_id);
+        }
+      }
+
+      for (int output_index = 0; output_index < op_kernel_context.OutputCount(); ++output_index) {
+        Fence_t fence = op_kernel_context.OutputFence(output_index);
+        if (fence) {
+          fence->BeforeUsingAsOutput(p_op_kernel->Node().GetExecutionProviderType(), queue_id);
+        }
+      }
+    }
```

Comment:
would be nice to avoid code copy by making it into util functions

Reply:
Copying code was done on purpose since @mrry had requested to minimize touching inference code as much as possible and then in a second PR consolidate/refactor code. I agree with you that we should be sharing code as much as possible, the question is do we do it in this PR or a seperate PR?